src/arithmetic.jl

```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
 
 
 ## Addition and substraction ##
 for f in (:+, :-)
-    for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
+    dotf = ifelse( f == :+, :(.+), :(.-))
+    for T in (:Taylor1, :TaylorN)
         @eval begin
             ($f){T<:Number,S<:Number}(a::$T{T}, b::$T{S}) = $f(promote(a,b)...)
             function $f{T<:Number}(a::$T{T}, b::$T{T})
-                a, b = fixorder(a, b)
-                v = similar(a.coeffs)
-                @simd for i in eachindex(v)
-                    @inbounds v[i] = $f(a.coeffs[i], b.coeffs[i])
+                aord = a.order
+                bord = b.order
+                ac = view(a.coeffs, :)
+                bc = view(b.coeffs, :)
+                if aord == bord
+                    v = similar(ac)
+                    v .= $dotf(ac, bc)
+                elseif aord < bord
+                    v = similar(bc)
+                    @inbounds v[1:aord+1] .= $dotf(ac[1:aord+1], bc[1:aord+1])
+                    @inbounds v[aord+2:bord+1] .= $f(bc[aord+2:bord+1])
+                else
+                    v = similar(ac)
+                    @inbounds v[1:bord+1] .= $dotf(ac[1:bord+1], bc[1:bord+1])
+                    @inbounds v[bord+2:aord+1] .= ac[bord+2:aord+1]
                 end
-                return $T(v, a.order)
+                return $T(v)
             end
             function $f(a::$T)
                 v = similar(a.coeffs)
-                @simd for i in eachindex(v)
-                    @inbounds v[i] = $f(a.coeffs[i])
-                end
+                broadcast!($f, v, a.coeffs)
                 return $T(v, a.order)
             end
-            if $T != HomogeneousPolynomial
-                function $f(a::$T, b::RealOrComplex)
-                    R = promote_type(eltype(a.coeffs), eltype(b))
-                    coeffs = Array{R}(a.order+1)
-                    @simd for i in eachindex(coeffs)
-                        @inbounds coeffs[i] = a.coeffs[i]
-                    end
-                    @inbounds coeffs[1] = $f(a.coeffs[1], b)
-                    return $T(coeffs, a.order)
-                end
-                function $f(b::RealOrComplex, a::$T)
-                    R = promote_type(eltype(a.coeffs), eltype(b))
-                    coeffs = Array{R}(a.order+1)
-                    @simd for i in eachindex(coeffs)
-                        @inbounds coeffs[i] = $f(a.coeffs[i])
-                    end
-                    @inbounds coeffs[1] = $f(b, a.coeffs[1])
-                    return $T(coeffs, a.order)
-                end
+            ($f){T<:Number,S<:Number}(a::$T{T}, b::S) = $f(promote(a,b)...)
+            function $f{T<:Number}(a::$T{T}, b::T)
+                R = eltype(a.coeffs)
+                coeffs = Array{R}(a.order+1)
+                coeffs .= a.coeffs
+                @inbounds coeffs[1] = $f(a.coeffs[1], b)
+                return $T(coeffs, a.order)
+            end
+            ($f){T<:Number,S<:Number}(b::S, a::$T{T}) = $f(promote(b,a)...)
```

Comment:
For `+` and `-` it's worth the code, since promotion would add lots of coefficients to zero.